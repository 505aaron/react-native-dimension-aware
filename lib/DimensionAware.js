import { Component } from "react";
import PropTypes from "prop-types";
import { Dimensions } from "react-native";
import reducer, { DIMENSION_CHANGE, DIMENSION_HYDRATE } from "./reducer";

export default class DimensionAware extends Component {
  static propTypes = {
    render: PropTypes.func.isRequired
  };

  constructor(props) {
    super(props);

    this.state = {
      data: reducer({}, { type: DIMENSION_HYDRATE })
    };
  }

  componentDidMount() {
    Dimensions.addEventListener("change", this.dimensionHandler);
  }

  componentWillUnmount() {
    Dimensions.removeEventListener("change", this.dimensionHandler);
  }

  dimensionHandler = ({ window, screen }) => {
    this.setState(({ data }) => ({
      data: reducer(data, { type: DIMENSION_CHANGE, window, screen })
    }));
  };

  render() {
    const { render } = this.props;
    const { data } = this.state;

    return render(data);
  }
}
