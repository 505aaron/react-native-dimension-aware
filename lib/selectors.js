export const getScreen = state => state.screen;
export const getScreenWidth = state => state.screen && state.screen.width;
export const getScreenHeight = state => state.screen && state.screen.height;

export const getWindow = state => state.window;
export const getWindowWidth = state => state.window && state.window.width;
export const getWindowHeight = state => state.window && state.window.height;
