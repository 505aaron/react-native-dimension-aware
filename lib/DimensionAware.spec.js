import React from "react";
import { View, Dimensions, Text } from "react-native";
import TestRenderer from "react-test-renderer";
import ShallowRenderer from "react-test-renderer/shallow";
import DimensionAware from "./DimensionAware";
import { getWindowWidth, getWindowHeight } from "./selectors";

describe("DimensionAware", () => {
  let shallowRenderer;
  let component;

  beforeEach(() => {
    shallowRenderer = new ShallowRenderer();

    component = (
      <DimensionAware
        render={dimensions => (
          <View>
            <Text>Width {getWindowWidth(dimensions)}</Text>
            <Text>Height {getWindowHeight(dimensions)}</Text>
          </View>
        )}
      />
    );
  });

  test("renders", () => {
    shallowRenderer.render(component);

    const tree = shallowRenderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  describe("lifecyle", () => {
    test("should re-render the component on updates", () => {
      const testRenderer = TestRenderer.create(component);

      Dimensions.set({ window: { width: 4000, height: 2000 } });

      testRenderer.update(component);

      expect(testRenderer.toJSON()).toMatchSnapshot();
    });

    test("should remove listeners", () => {
      const testRenderer = TestRenderer.create(component);

      testRenderer.unmount();

      Dimensions.set({ window: { width: 4000, height: 2000 } });

      expect(testRenderer.toJSON()).toBeNull();
    });
  });
});
