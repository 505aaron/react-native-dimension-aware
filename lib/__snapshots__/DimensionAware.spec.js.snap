// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DimensionAware lifecyle screen should override both 1`] = `
<View>
  <Text>
    Width 
    60
  </Text>
  <Text>
    Height 
    30
  </Text>
</View>
`;

exports[`DimensionAware lifecyle screen should override height 1`] = `
<View>
  <Text>
    Width 
    4000
  </Text>
  <Text>
    Height 
    30
  </Text>
</View>
`;

exports[`DimensionAware lifecyle screen should override width 1`] = `
<View>
  <Text>
    Width 
    30
  </Text>
  <Text>
    Height 
    2000
  </Text>
</View>
`;

exports[`DimensionAware lifecyle screen should re-render the component on updates 1`] = `
<View>
  <Text>
    Width 
    4000
  </Text>
  <Text>
    Height 
    2000
  </Text>
</View>
`;

exports[`DimensionAware lifecyle window should override both 1`] = `
<View>
  <Text>
    Width 
    60
  </Text>
  <Text>
    Height 
    30
  </Text>
</View>
`;

exports[`DimensionAware lifecyle window should override height 1`] = `
<View>
  <Text>
    Width 
    4000
  </Text>
  <Text>
    Height 
    30
  </Text>
</View>
`;

exports[`DimensionAware lifecyle window should override width 1`] = `
<View>
  <Text>
    Width 
    30
  </Text>
  <Text>
    Height 
    2000
  </Text>
</View>
`;

exports[`DimensionAware lifecyle window should re-render the component on updates 1`] = `
<View>
  <Text>
    Width 
    4000
  </Text>
  <Text>
    Height 
    2000
  </Text>
</View>
`;

exports[`DimensionAware renders 1`] = `
<Component>
  <Component>
    Width 
    750
  </Component>
  <Component>
    Height 
    1334
  </Component>
</Component>
`;
